src/pf_driver/src/pf/timesync.cpp

```diff
@@ -0,0 +1,315 @@
+long TimeSync::time_to_full_sensor_second(rclcpp::Time& pc_time)
+{
+  std::lock_guard<std::mutex> guard(access_);
+
+  std::chrono::duration<double> conv_time(((pc_time - pc_base_).seconds() - off_usec_ - base_time_) /
+                                          (1.0 + scale_time_));
+
+  rclcpp::Time sensor_time = sensor_base_ + rclcpp::Duration(conv_time);
+
+  /* Return remainder until full second on sensor */
+  return 1000000000l - (sensor_time.nanoseconds() % 1000000000);
+}
+
+void TimeSync::update(uint64_t sensor_time_raw, unsigned req_duration_us, rclcpp::Time pc_time)
+{
+  TimeSync_Sample sample;
+
+  if (req_duration_us > 50000)
+  {
+    return;
+  }
+
+  if (!samples_.empty())
+  {
+    /* Reset stats if there was a "large" jump in pc time, e.g. due to daylight savings?? */
+    /* TBD: What is "large"? 15sec? 15min? */
+
+    double since_last_update = (pc_time - samples_.back().pc_time).seconds();
+    if (fabs(since_last_update) > 15.0)
+    {
+      reset(since_last_update);
+    }
+  }
+
+  {
+    double period_secs = 0.001 * (double)period_;
+
+    raw_to_rclcpp(sensor_time_raw, sample.sensor_time, pc_time.get_clock_type());
+
+    if (!samples_.empty())
+    {
+      double since_last_update = (sample.sensor_time - samples_.back().sensor_time).seconds();
+
+      if (fabs(since_last_update) > 1.0)
+      {
+        /* Reset stats if there was a long time since last packet, probably some kind of restart */
+
+        reset(since_last_update);
+      }
+      else
+      {
+        /* Do not blindly collect all samples, as that might cause the
+         * collection to grow very large if user configured to observe input
+         * over a long time at high scan rate.  Rather ignore new samples that
+         * come in at a higher rate than necessary to collect just 200 samples
+         * evenly distributed during the whole period.  TBD: 200 was chosen
+         * somewhat arbitrarily.
+         */
+
+        if (since_last_update < period_secs / 200)
+        {
+          return;
+        }
+      }
+    }
+
+    /* Discard samples older than period_ */
+    while (!samples_.empty() && ((sample.sensor_time - samples_.front().sensor_time).seconds()) > period_secs)
+    {
+      sum_req_duration_us_ -= samples_.front().req_duration_us;
+      samples_.pop_front();
+    }
+
+    /* Record new sample */
+    sample.req_duration_us = req_duration_us;
+    sum_req_duration_us_ += req_duration_us;
+    sample.pc_time = pc_time;
+    samples_.push_back(sample);
+  }
+
+  /* Offsets to keep values small in calculation below */
+
+  /* These have to be considered in addition to the computed
+    offset and scale, see sensor_to_pc() method above */
+
+  rclcpp::Time sensor_base(samples_.back().sensor_time);
+  rclcpp::Time pc_base(samples_.back().pc_time);
+
+  {
+    double time_factor = 1.0;
+    double base_time = 0.0;
+
+    if (averaging_ = TIMESYNC_AVERAGING_REGRESSION)
```

Comment:
Evil trap...  And I wonder why averaging_ is 0 in the first run and then always 1.
```suggestion
    if (TIMESYNC_AVERAGING_REGRESSION == averaging_)
```